File: src/test/java/com/loc/microservices/order_service_spring_boot/controller/OrderControllerTest.java
package com.loc.microservices.order_service_spring_boot.controller;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.loc.microservices.order_service_spring_boot.dto.OrderRequest;
import com.loc.microservices.order_service_spring_boot.service.OrderService;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest;
import org.springframework.http.MediaType;
import org.springframework.test.context.bean.override.mockito.MockitoBean;
import org.springframework.test.web.servlet.MockMvc;

import java.math.BigDecimal;

import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.doNothing;
import static org.mockito.Mockito.doThrow;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.post;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.content;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.status;

@WebMvcTest(OrderController.class)
public class OrderControllerTest {

    @Autowired
    private MockMvc mockMvc;
    @MockitoBean
    private OrderService orderService;
    @MockitoBean
    private ObjectMapper objectMapper;

    @Test
    void placeOrder_withValidRequest_shouldReturnSuccess() throws Exception {
        // Arrange
        OrderRequest orderRequest = new OrderRequest(
            null, // id
            null, // orderNumber
            "test-product", // skuCode
            BigDecimal.valueOf(100.0), // price
            1 // quantity
        );
        
        doNothing().when(orderService).placeOrder(any(OrderRequest.class));

        // Act & Assert
        mockMvc.perform(post("/api/order")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(orderRequest)))
                .andExpect(status().isOk())
                .andExpect(content().string("Order placed successfully"));
    }

    @Test
    void placeOrder_whenProductNotInStock_shouldReturnError() throws Exception {
        // Arrange
        OrderRequest orderRequest = new OrderRequest(
            null, // id
            null, // orderNumber
            "out-of-stock", // skuCode
            BigDecimal.valueOf(100.0), // price
            1 // quantity
        );
        
        doThrow(new RuntimeException("Product with Skucode: out-of-stock is not in stock"))
                .when(orderService).placeOrder(any(OrderRequest.class));

        // Act & Assert
        mockMvc.perform(post("/api/order")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(orderRequest)))
                .andExpect(status().isInternalServerError());
    }
} 

File: src/test/kotlin/com/loc/demo_microservices/controller/OrderControllerTest.kt
package com.loc.order_service.controller

import com.fasterxml.jackson.databind.ObjectMapper
import com.loc.order_service.model.Order
import com.loc.order_service.service.OrderService
import com.loc.orderservice.model.OrderRequest
import io.mockk.junit5.MockKExtension
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.WebMvcTest
import org.springframework.test.web.servlet.MockMvc
import com.ninjasquad.springmockk.MockkBean
import io.mockk.every
import io.mockk.impl.annotations.MockK
import io.restassured.http.ContentType
import io.restassured.module.mockmvc.RestAssuredMockMvc
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import io.restassured.module.mockmvc.kotlin.extensions.Given
import io.restassured.module.mockmvc.kotlin.extensions.Then
import io.restassured.module.mockmvc.kotlin.extensions.When
import org.springframework.http.HttpStatus
import java.math.BigDecimal
import java.time.LocalDateTime

@WebMvcTest(OrderController::class)
@ExtendWith(MockKExtension::class)
internal class OrderControllerTest (
    @Autowired private val mockMvc: MockMvc,
    @Autowired private val objectMapper: ObjectMapper
) {
    @MockK
    private lateinit var orderService: OrderService

    @BeforeEach
    fun setup() {
        RestAssuredMockMvc.mockMvc(mockMvc)
    }

    @Test
    fun `should return CREATED when creating a new order`() {
        // Given
        val orderRequest = OrderRequest(
            skuCode = "123",
            price = BigDecimal("999.99"),
            quantity = 2
        )

        val createdOrder = Order(
            id = 1L,
            orderNumber = "ORD-1234567890-ABCD1234",
            skuCode = "IPHONE13-128",
            price = BigDecimal("999.99"),
            quantity = 2,
            createdAt = LocalDateTime.now()
        )

        every { orderService.createOrder(any()) } returns createdOrder

        //When / Then
        Given {
            contentType(ContentType.JSON)
            body(objectMapper.writeValueAsString(orderRequest))
        } When  {
            post("/api/orders")
        } Then  {
            status(HttpStatus.CREATED)
        }
    }
}